src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -328,7 +329,9 @@ private string GenerateId()
                 //sanitize external RequestId as it may not be hierarchical. 
                 //we cannot update ParentId, we must let it be logged exactly as it was passed.
                 string parentId = ParentId[0] == RootIdPrefix ? ParentId : RootIdPrefix + ParentId;
-                if (parentId[parentId.Length - 1] != '.')
+
+                char lastChar = parentId[parentId.Length - 1];
+                if (lastChar != '.' && lastChar != '_')
```

Comment:
This is pure readability thing.  Nothing must break if there are two consecutive delimiters (which means empty node) or if there are two consecutive nodes with '_' at the end.

We even have a [standard proposal](https://github.com/lmolkova/correlation/blob/master/hierarchical_request_id.md) for the Id: that does not specify behavior in this case.